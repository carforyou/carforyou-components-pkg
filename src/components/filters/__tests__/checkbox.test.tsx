import React from "react"
import { fireEvent, render } from "@testing-library/react"

import CheckboxFilter from "../checkbox"
import ClockOutlinedTealIcon from "../../../assets/src/icons/clockOutlinedTeal"

const renderWrapper = ({
  name = "testFilter",
  options = [
    { label: "One", value: 1, renderIcon: null },
    { label: "Two", value: 2, renderIcon: null },
    { label: "Three", value: 3, renderIcon: null },
  ],
  applyFilters = jest.fn(),
  selected = [],
  facet = {},
  children = null,
  onSelect = jest.fn(),
} = {}) => {
  return render(
    <CheckboxFilter
      name={name}
      options={options}
      apply={applyFilters}
      selected={selected}
      facet={facet}
      onSelect={onSelect}
    >
      {children}
    </CheckboxFilter>
  )
}

describe("<CheckboxFilter/>", () => {
  describe("renders correctly", () => {
    it("with no options", () => {
      const { container } = renderWrapper({ options: [] })

      expect(container).toMatchSnapshot()
    })

    it("with simple options", () => {
      const { container } = renderWrapper({
        options: [
          { label: "One", value: 1, renderIcon: null },
          { label: "Two", value: 2, renderIcon: null },
          { label: "Three", value: 3, renderIcon: null },
        ],
      })

      expect(container).toMatchSnapshot()
    })

    it("with icons", () => {
      const { getByTestId } = renderWrapper({
        options: [
          {
            label: "One",
            value: 1,
            renderIcon: () => (
              <ClockOutlinedTealIcon
                width="20"
                height="20"
                className="mr-10"
                data-testid="icon-1"
              />
            ),
          },
          {
            label: "Two",
            value: 2,
            renderIcon: () => (
              <ClockOutlinedTealIcon width="20" height="20" className="mr-10" />
            ),
          },
          {
            label: "Three",
            value: 3,
            renderIcon: () => (
              <ClockOutlinedTealIcon width="20" height="20" className="mr-10" />
            ),
          },
        ],
      })

      expect(getByTestId("icon-1"))
    })

    it("with initial selection", () => {
      const { container } = renderWrapper({ selected: [1, 2] })

      expect(container).toMatchSnapshot()
    })

    it("with facet", () => {
      const { container } = renderWrapper({
        facet: { "1": 30, "2": 1000, "3": 0 },
      })

      expect(container).toMatchSnapshot()
    })
  })

  it("applies filter on click", () => {
    const applyFilterMock = jest.fn()
    const { getByText } = renderWrapper({ applyFilters: applyFilterMock })

    fireEvent.click(getByText("One"))

    expect(applyFilterMock).toHaveBeenCalledWith({ testFilter: [1] })
  })

  it("disables non-selected option with facet = zero", () => {
    const { getByLabelText } = renderWrapper({
      facet: { "1": 30, "2": 1000, "3": 0 },
    })

    expect(getByLabelText("Three", { exact: false })).toHaveProperty("disabled")
  })

  it("disables non-selected option when facet is provided and null", () => {
    const { getByLabelText } = renderWrapper({
      facet: { "1": 30, "2": 1000 },
    })

    expect(getByLabelText("Three", { exact: false })).toHaveProperty("disabled")
  })

  it("calls onSelect when the value is selected", () => {
    const onSelectFilterMock = jest.fn()
    const { getByText } = renderWrapper({ onSelect: onSelectFilterMock })

    fireEvent.click(getByText("Three"))

    expect(onSelectFilterMock).toHaveBeenCalledWith(3)
  })
})
