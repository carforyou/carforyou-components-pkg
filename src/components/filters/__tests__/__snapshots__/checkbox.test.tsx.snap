// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CheckboxFilter/> renders correctly with facet 1`] = `
<div>
  <div
    class="flex flex-wrap mb-5"
  >
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_1"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  class=""
                  data-valid="true"
                  id="filter_testFilter_1"
                  name="filter_testFilter_1"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  One
                  <span
                    class="inline-block mr-5 text-grey-3 text-sm"
                  >
                    (
                    30
                    )
                  </span>
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_2"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  class=""
                  data-valid="true"
                  id="filter_testFilter_2"
                  name="filter_testFilter_2"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  Two
                  <span
                    class="inline-block mr-5 text-grey-3 text-sm"
                  >
                    (
                    1â€™000
                    )
                  </span>
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-not-allowed text-grey-4 focus-within:text-teal"
            for="filter_testFilter_3"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  class=""
                  data-valid="true"
                  disabled=""
                  id="filter_testFilter_3"
                  name="filter_testFilter_3"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  Three
                  <span
                    class="inline-block mr-5 text-grey-3 text-sm"
                  >
                    (
                    0
                    )
                  </span>
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<CheckboxFilter/> renders correctly with initial selection 1`] = `
<div>
  <div
    class="flex flex-wrap mb-5"
  >
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_1"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  checked=""
                  class=""
                  data-valid="true"
                  id="filter_testFilter_1"
                  name="filter_testFilter_1"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  One
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_2"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  checked=""
                  class=""
                  data-valid="true"
                  id="filter_testFilter_2"
                  name="filter_testFilter_2"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  Two
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_3"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  class=""
                  data-valid="true"
                  id="filter_testFilter_3"
                  name="filter_testFilter_3"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  Three
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
  </div>
</div>
`;

exports[`<CheckboxFilter/> renders correctly with no options 1`] = `
<div>
  <div
    class="flex flex-wrap mb-5"
  />
</div>
`;

exports[`<CheckboxFilter/> renders correctly with simple options 1`] = `
<div>
  <div
    class="flex flex-wrap mb-5"
  >
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_1"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  class=""
                  data-valid="true"
                  id="filter_testFilter_1"
                  name="filter_testFilter_1"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  One
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_2"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  class=""
                  data-valid="true"
                  id="filter_testFilter_2"
                  name="filter_testFilter_2"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  Two
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
    <div
      class="w-12/12 -mb-5"
    >
      <div
        class="relative pb-px"
      >
        <div
          class="validation-inputContainer mb-20"
        >
          <label
            class="block w-12/12 cursor-pointer focus-within:text-teal"
            for="filter_testFilter_3"
          >
            <div
              class="flex leading-sm font-normal items-center justify-between flex-row-reverse"
            >
              <span
                class="flex items-center ml-5"
              >
                <input
                  class=""
                  data-valid="true"
                  id="filter_testFilter_3"
                  name="filter_testFilter_3"
                  type="checkbox"
                  value=""
                />
              </span>
              <span
                class="w-12/12"
              >
                <span
                  class="text-base leading-xs flex items-center justify-between"
                >
                  Three
                </span>
              </span>
            </div>
          </label>
        </div>
        <div />
      </div>
    </div>
  </div>
</div>
`;
